[BUG] Can not ansi cast decimal type to long type while fetching decimal column from data table

**Describe the bug**
Cudf Exception thrown when trying to cast decimal to long

**Steps/Code to reproduce bug**
```
spark.conf.set("spark.sql.ansi.enabled", true)
val df = Seq(222.22, 777.77).toDF("f")
val df2 = df.repartition(1).selectExpr("cast(f as decimal(5,2))")
val df3 = df2.repartition(1).selectExpr("cast(f as long)")
df3.show
```

```
Caused by: ai.rapids.cudf.CudfException: cuDF failure at: /home/jenkins/agent/workspace/jenkins-spark-rapids-jni_nightly-pre_release-29-cuda11/thirdparty/cudf/cpp/src/binaryop/binaryop.cpp:205: Unsupported operator for these types
  at ai.rapids.cudf.ColumnView.binaryOpVS(Native Method)
  at ai.rapids.cudf.ColumnView.binaryOp(ColumnView.java:1254)
  at ai.rapids.cudf.ColumnView.binaryOp(ColumnView.java:1244)
  at ai.rapids.cudf.BinaryOperable.lessThan(BinaryOperable.java:277)
  at ai.rapids.cudf.BinaryOperable.lessThan(BinaryOperable.java:284)
  at com.nvidia.spark.rapids.GpuCast$.$anonfun$assertValuesInRange$3(GpuCast.scala:641)
```
**Expected behavior**
The script should not throw an exception

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
@@ -378,17 +378,17 @@ object GpuCast extends Arm {
       // ansi cast from larger-than-long integral-like types, to long
       case (dt: DecimalType, LongType) if ansiMode =>
         // This is a work around for https://github.com/rapidsai/cudf/issues/9282
-        val bigDecimalMin = BigDecimal(Long.MinValue)
+        val min = BigDecimal(Long.MinValue)
             .setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
-        val bigDecimalMax = BigDecimal(Long.MaxValue)
+        val max = BigDecimal(Long.MaxValue)
             .setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
         // We are going against our convention of calling assertValuesInRange()
         // because the min/max values are a different decimal type i.e. Decimal 128 as opposed to
         // the incoming input column type.
-        withResource(input.min()) { min =>
-          withResource(input.max()) { max =>
-            if (min.isValid && min.getBigDecimal().compareTo(bigDecimalMin) == -1 ||
-                max.isValid && max.getBigDecimal().compareTo(bigDecimalMax) == 1) {
+        withResource(input.min()) { minInput =>
+          withResource(input.max()) { maxInput =>
+            if (minInput.isValid && minInput.getBigDecimal().compareTo(min) == -1 ||
+                maxInput.isValid && maxInput.getBigDecimal().compareTo(max) == 1) {
               throw new ArithmeticException(GpuCast.INVALID_INPUT_MESSAGE)
             }
           }

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
@@ -387,8 +387,8 @@ object GpuCast extends Arm {
         // the incoming input column type.
         withResource(input.min()) { min =>
           withResource(input.max()) { max =>
-            if (min.getBigDecimal().compareTo(bigDecimalMin) == -1 ||
-                max.getBigDecimal().compareTo(bigDecimalMax) == 1) {
+            if (min.isValid && min.getBigDecimal().compareTo(bigDecimalMin) == -1 ||
+                max.isValid && max.getBigDecimal().compareTo(bigDecimalMax) == 1) {
               throw new ArithmeticException(GpuCast.INVALID_INPUT_MESSAGE)
             }
           }

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
@@ -386,7 +386,7 @@ object GpuCast extends Arm {
           withResource(input.max()) { max =>
             if (min.getBigDecimal().compareTo(bigDecimalMin) == -1 ||
                 max.getBigDecimal().compareTo(bigDecimalMax) == 1) {
-              throw new IllegalStateException(GpuCast.INVALID_INPUT_MESSAGE)
+              throw new ArithmeticException(GpuCast.INVALID_INPUT_MESSAGE)
             }
           }
         }

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
@@ -177,6 +177,8 @@ object GpuCast extends Arm {
   private val TIMESTAMP_TRUNCATE_REGEX = "^([0-9]{4}-[0-9]{2}-[0-9]{2} " +
     "[0-9]{2}:[0-9]{2}:[0-9]{2})" +
     "(.[1-9]*(?:0)?[1-9]+)?(.0*[1-9]+)?(?:.0*)?$"
+  private val bigDecimalLongMin = BigDecimal(Long.MinValue)
+  private val bigDecimalLongMax = BigDecimal(Long.MaxValue)
 
   val INVALID_INPUT_MESSAGE: String = "Column contains at least one value that is not in the " +
     "required range"
@@ -378,10 +380,8 @@ object GpuCast extends Arm {
       // ansi cast from larger-than-long integral-like types, to long
       case (dt: DecimalType, LongType) if ansiMode =>
         // This is a work around for https://github.com/rapidsai/cudf/issues/9282
-        val min = BigDecimal(Long.MinValue)
-            .setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
-        val max = BigDecimal(Long.MaxValue)
-            .setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
+        val min = bigDecimalLongMin.setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
+        val max = bigDecimalLongMax.setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
         // We are going against our convention of calling assertValuesInRange()
         // because the min/max values are a different decimal type i.e. Decimal 128 as opposed to
         // the incoming input column type.

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
@@ -177,8 +177,8 @@ object GpuCast extends Arm {
   private val TIMESTAMP_TRUNCATE_REGEX = "^([0-9]{4}-[0-9]{2}-[0-9]{2} " +
     "[0-9]{2}:[0-9]{2}:[0-9]{2})" +
     "(.[1-9]*(?:0)?[1-9]+)?(.0*[1-9]+)?(?:.0*)?$"
-  private val bigDecimalLongMin = BigDecimal(Long.MinValue)
-  private val bigDecimalLongMax = BigDecimal(Long.MaxValue)
+  private val BIG_DECIMAL_LONG_MIN = BigDecimal(Long.MinValue)
+  private val BIG_DECIMAL_LONG_MAX = BigDecimal(Long.MaxValue)
 
   val INVALID_INPUT_MESSAGE: String = "Column contains at least one value that is not in the " +
     "required range"
@@ -380,8 +380,8 @@ object GpuCast extends Arm {
       // ansi cast from larger-than-long integral-like types, to long
       case (dt: DecimalType, LongType) if ansiMode =>
         // This is a work around for https://github.com/rapidsai/cudf/issues/9282
-        val min = bigDecimalLongMin.setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
-        val max = bigDecimalLongMax.setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
+        val min = BIG_DECIMAL_LONG_MIN.setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
+        val max = BIG_DECIMAL_LONG_MAX.setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
         // We are going against our convention of calling assertValuesInRange()
         // because the min/max values are a different decimal type i.e. Decimal 128 as opposed to
         // the incoming input column type.

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
@@ -382,6 +382,9 @@ object GpuCast extends Arm {
             .setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
         val bigDecimalMax = BigDecimal(Long.MaxValue)
             .setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
+        // We are going against our convention of calling assertValuesInRange()
+        // because the min/max values are a different decimal type i.e. Decimal 128 as opposed to
+        // the incoming input column type.
         withResource(input.min()) { min =>
           withResource(input.max()) { max =>
             if (min.getBigDecimal().compareTo(bigDecimalMin) == -1 ||

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala
@@ -378,11 +378,18 @@ object GpuCast extends Arm {
       // ansi cast from larger-than-long integral-like types, to long
       case (dt: DecimalType, LongType) if ansiMode =>
         // This is a work around for https://github.com/rapidsai/cudf/issues/9282
-        val min = BigDecimal(Long.MinValue)
+        val bigDecimalMin = BigDecimal(Long.MinValue)
             .setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
-        val max = BigDecimal(Long.MaxValue)
+        val bigDecimalMax = BigDecimal(Long.MaxValue)
             .setScale(dt.scale, BigDecimal.RoundingMode.DOWN).bigDecimal
-        assertValuesInRange(input, Scalar.fromDecimal(min), Scalar.fromDecimal(max))
+        withResource(input.min()) { min =>
+          withResource(input.max()) { max =>
+            if (min.getBigDecimal().compareTo(bigDecimalMin) == -1 ||
+                max.getBigDecimal().compareTo(bigDecimalMax) == 1) {
+              throw new IllegalStateException(GpuCast.INVALID_INPUT_MESSAGE)
+            }
+          }
+        }
         if (dt.precision <= DType.DECIMAL32_MAX_PRECISION && dt.scale < 0) {
           // This is a work around for https://github.com/rapidsai/cudf/issues/9281
           withResource(input.castTo(DType.create(DType.DTypeEnum.DECIMAL64, -dt.scale))) { tmp =>

--- a/tests/src/test/scala/com/nvidia/spark/rapids/AnsiCastOpSuite.scala
+++ b/tests/src/test/scala/com/nvidia/spark/rapids/AnsiCastOpSuite.scala
@@ -466,6 +466,11 @@ class AnsiCastOpSuite extends GpuExpressionTestSuite {
     }
   }
 
+  testSparkResultsAreEqual("ansi_cast decimals to long",
+    generateValidValuesDecimalDF(Short.MinValue, Short.MaxValue, 18, 3), sparkConf) {
+    frame => testCastTo(DataTypes.LongType)(frame)
+  }
+
   private def castToStringExpectedFun[T]: T => Option[String] = (d: T) => Some(String.valueOf(d))
 
   private def testCastToString[T](dataType: DataType, ansiMode: Boolean,

